### ISSUE TITLE
Support building directory/glob of SASS files from CLI

### ISSUE THREAD

**Issue by patrickarlt** [sass/node-sass#828]
Coming from Ruby SASS the one feature I am missing with the node-sass command line is the ability to build a directory of SASS files. For example in Ruby SASS I can do...

```
$ sass --watch sass
>>> Sass is watching for changes. Press Ctrl-C to stop.
      write sass/one.css
      write sass/one.css.map
      write sass/two.css
      write sass/two.css.map
```

And a build is kicked off immediately and it watches and rebuilds. But the same in Node SASS produces a usage message.

```
$ node-sass --watch sass
Provide a Sass file to render

  Example
    node-sass --output-style compressed foobar.scss foobar.css
    cat foobar.scss | node-sass --output-style compressed > foobar.css
```
This is really needed for parity with Ruby SASS and for larger projects where they might be multiple separate files that need to be built.

**Comment by paulcpederson** [comment]
for now, you can use [scss-cli](https://github.com/paulcpederson/scss-cli)

**Comment by am11** [comment]
@patrickarlt, thanks for your feedback, we will have a look. Albeit, this might not land in v3.0 but we have this planned.
@paulcpederson, would you like to send a PR to node-sass instead?

/cc @wesleytodd

**Comment by paulcpederson** [comment]
@am11 yeah! I can definitely try! It would be much better to have this behavior in node-sass, always intended for scss-cli to be a band-aid measure, for sure.
 
Looking at #685 it looks like you started with this, but it was closed. What happened?

Also, what @patrickarlt is expecting, which I think is pretty common, is that if node-sass is pointed at a directory, it will recursively compile all sass files in said directory *except* for sass files beginning with an `_`. Obviously you could build this functionality yourself using grep or bash scripts, but IMO this is well within the scope of node-sass. So the feature requests are:

1. Compile a directory of sass files to css
2. Ignore sass files that begin with `_`

Currently in ruby sass this works with the `--watch` flag, but it also works with the (undocumented?) `--update` flag. So I can do something like `sass --update stylesheets:css` and it will recursively compile sass files in `stylesheets/` to `css/`. I've never been a huge fan of this syntax as it feels counter intuitive to use a flag like `--update` when aren't I always updating things when I compile *any* sass?

To my mind, there are a couple ways to accomplish this in node-sass: 

1. Using globs like `node-sass sass/**/*.scss -o css/ `
2. Automatically detecting a directory, like `node-sass sass/ -o css/`
3. Creating a new flag to explicitly use a directory like  `node-sass sass/ -d -o css/` (`-d, --directory`)

Not sure which would be best, or if I'm way off base. Interested to hear people's thoughts.

**Comment by am11** [comment]
> but it was closed. What happened?

This: https://github.com/sass/node-sass/issues/601#issuecomment-75373433

**Comment by paulcpederson** [comment]
Ok, piecing together the history. 

> multi-file compilation support, cannot be entertained.

So you consider this out of scope *all together*?

>Nonetheless, I tried implementing multi-file compilation with globs and no matter how we slice it, there are blockers like:
>(Considering glob can accept patterns like: {**/styles/*, /totally/different/path/{*.sass, *.scss}, /blah/**})

Seems like globs get into very hairy territory. That would certainly eliminate option 1.

**Comment by am11** [comment]
> So you consider this out of scope all together?

Nope, given the scope of #601, the OP's concern was to fix behavior of `--recursive`, which was eventually addressed by @wesleytodd (via #782). The multiple compilation was beyond the scope of #601. I opened the experimental PR (#685) to enhance various aspects of CLI, but it turned out to be a hard to determine and need use-case driven gradual improvement.

From the experience of #685, the real challenge was to decide what kind of structure the output directory will follow, keeping in mind that globs support uniting different paths/drives. Perhaps this discussion will yield solid solution and we will eventually end up with this ability.

For instance:

Input:  `{/styles/, /totally/different/path/{.sass, *.scss}, /blah/})`
Output directory: /home/user-name/projects/sass-practice/

Question: what should be the structure of output directory, given the input glob paths have multiple nested directories?

IMO, we can take ideas from other node-based compilers (coffeescript, icedcoffeescript, less, stylus, livescript, sweetjs, typescript etc.) and find out how they are handling glob for inputs (if any of them support it).

**Comment by wesleytodd** [comment]
My 2c on this:

When compiling a directory, you can only specify one.  If you want to watch more than one directory you have to run multiple processes.  Maybe I'm not seeing it, but it appears that right now `node-sass` only accepts one input currently.  Which would mean nothing there changes.

With that restriction solving for @am11's example, the answer becomes fairly simple.  Mirror the directory structure inside the input directory into the output directory.

For example: 

`$ node-sass sass build`

Given this structure:

```
- sass/
| - main.scss  # Includes components dir
|  | - components/
|  |  | - _component1.scss
|  | - pages/
|  |  | - page1.scss
|  |  | - page2.scss
```

You would get this output:

```
- sass/
- build/
| - main.css
|  | - pages/
|  |  | - page1.css
|  |  | - page2.css
```

Is this in line with what everyone else was thinking?

**Comment by paulcpederson** [comment]
+1 to only accepting 1 directory. This is the most common use case.

@wesleytodd what are your thoughts about the best way to accomplish that?

**Comment by patrickarlt** [comment]
> Mirror the directory structure inside the input directory into the output directory.

@wesleytodd @paulcpederson This is exactly what I would want and expect. Globs (especially multiple globs) seems like it would be way to difficult. If you really need something like that you should definitely turn to something like Gulp or Grunt.

But building all the `.scss` files in a single directory and mirroring the structure to the output dir would be awesome.

For implementation it would be fairly simple to do something like this.

1. Match everything in the directory with something like [node-glob](https://github.com/isaacs/node-glob)
2. Build the SASS for everything in the directory that does not start with `_`
3. When you need to build the output files just replace the path to the directory `/path/to/sass` with an empty string. this would give you something like `main.scss` and `pages/page1.scss`.
4. `path.dirname` (https://nodejs.org/api/path.html#path_path_dirname_p) the output filename. If you need for build a deeply nested directory structure you can use https://github.com/substack/node-mkdirp
5. Write the output files into the newly created directories.

**Comment by paulcpederson** [comment]
@patrickarlt yeah I know all that, I guess I meant more from an interface perspective where does multiple file compilation fit in.  Is it automatically detected by the cli? Or is it a new flag `-d, --directory`? Or something else?

**Comment by am11** [comment]
@wesleytodd, :+1: for the single directory approach and that multi-path-union globs would be a mess.

@patrickarlt,

>  Is it automatically detected by the cli?

Yes, that would make perfect sense. Additionally:

* To specify input directory and output directory:
  ```
  node-sass sass/  -o build/
  ```

* Error if value of -o [is not a directory](https://github.com/sass/node-sass/pull/685/files#diff-c19ab2481e2703fcf0d2808f4c762f63R204).


* If input is directory and output is absent, error:
  >
  Multi-file compile mode: cannot print to stdout. Please specify directory.

  (since missing output path now [prints to stdout](https://github.com/sass/node-sass/commit/433e939d8ecb84c1d0deed4dbb055780b99de331))

**Comment by paulcpederson** [comment]
@am11 ok cool! I have a first take at this running locally. 

Let me add some of the errors you mention above and clean up a bit and then I can try to get something up for review.

**Comment by wesleytodd** [comment]
One other note, we should make sure this use case still respects the `--recursive` flag.  Probably have a specific test case for it.  This seems like the most common place where you might specifically want, or not want, recursive.

**Comment by paulcpederson** [comment]
@wesleytodd to clarify, you're saying that if somebody runs `node-sass sass/ -o css/` then by default it is **not** recursive?

**Comment by am11** [comment]
@patrickarlt, it is recursive, but if someone specify `node-sass --recursive false ..` then it is not.
On this note, we should replace `--recursive` with `--no-recursive`, since it is must better than saying `--recursive false` IMO.

**Comment by wesleytodd** [comment]
@am11 

> On this note, we should replace --recursive with --no-recursive, since it is must better than saying --recursive false IMO.

Completely agree!  Much better standard.

**Comment by paulcpederson** [comment]
+1 on `no-recursive`

**Comment by paulcpederson** [comment]
Ok, got automatic multiple file compilation working with `node-sass sass/ -o css/`. Now I'm trying to write some tests. I ran the tests, and some failed, which I assumed meant I broke something. But after checking out master and running the tests from there, I think they might just be broken? Which is weird because Travis says the build is all good. Maybe I'm missing something about setting up the test suite?

The failing tests are:

```
1) api .render(options, callback) should compile generate map with sourceMapRoot pass-through option:
     Uncaught AssertionError: undefined == 'http://test.com/'

  2) api .render(importer) should reflect user-defined error when returned as callback:
     Uncaught TypeError: Cannot read property 'message' of null

  3) api .render(importer) should reflect user-defined error with return:
     Uncaught TypeError: Cannot read property 'message' of null

  4) api .renderSync(options) should compile generate map with sourceMapRoot pass-through option:
     AssertionError: undefined == 'http://test.com/'

  5) api .renderSync(importer) should throw user-defined error:
     AssertionError: Missing expected exception..

  6) cli node-sass in.scss --output out.css should compile with the --source-root option:
     Uncaught AssertionError: undefined == 'http://test/'

  7) cli importer should reflect user-defined Error:
     Uncaught SyntaxError: Unexpected token R
```

**Comment by am11** [comment]
@paulcpederson, you would need to rebuild the binaries.

I use this kind of a workflow:

```bash
# non-recurring 
cd ~/projects/github
git clone https://github.com/paulcpederson/node-sass --recursive
cd node-sass
git remote add node-sass https://github.com/sass/node-sass
git pull --rebase node-sass master
git submodule --init --recursive
```

```bash
# recurring 
cd ~/projects/github
git pull --rebase node-sass master
node scripts/build -f   ## <-- this will rebuild the binaries
npm test
node

# node.js interactive console
require('./').render({data: 'what{ev:er}'}, function(err, result){console.log(err, result)})
# and so forth
```

**Comment by paulcpederson** [comment]
@am11 beautiful thank you I'll give that a shot.

**Comment by paulcpederson** [comment]
Good news: that worked great. 
Bad news: I (of course) have failed a couple tests. 

Let me see if I can get everything passing again and then I can put up something for review.

**Comment by am11** [comment]
This is addressed by 3d44ce88d5fcefbf70abb9e8b69eb98d06ffcc16 via #838. Credit: @paulcpederson. :tada:
